pkg/webhook/mutation.go

```diff
@@ -0,0 +1,168 @@
+// +kubebuilder:rbac:groups=*,resources=*,verbs=get;list;watch;update
+
+// AddMutatingWebhook registers the mutating webhook server with the manager
+func AddMutatingWebhook(mgr manager.Manager, client *opa.Client, processExcluder *process.Excluder) error {
+	if !*mutation.MutationEnabled {
+		return nil
+	}
+	reporter, err := newStatsReporter()
+	if err != nil {
+		return err
+	}
+	eventBroadcaster := record.NewBroadcaster()
+	kubeClient := kubernetes.NewForConfigOrDie(mgr.GetConfig())
+
+	eventBroadcaster.StartRecordingToSink(&clientcorev1.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
+	recorder := eventBroadcaster.NewRecorder(
+		scheme.Scheme,
+		corev1.EventSource{Component: "gatekeeper-mutation-webhook"})
+
+	wh := &admission.Webhook{
+		Handler: &mutationHandler{
+			webhookHandler: webhookHandler{
+				client:          mgr.GetClient(),
+				reader:          mgr.GetAPIReader(),
+				reporter:        reporter,
+				processExcluder: processExcluder,
+				eventRecorder:   recorder,
+				gkNamespace:     util.GetNamespace(),
+			},
+		},
+	}
+
+	// TODO(https://github.com/open-policy-agent/gatekeeper/issues/661): remove log injection if the race condition in the cited bug is eliminated.
+	// Otherwise we risk having unstable logger names for the webhook.
+	if err := wh.InjectLogger(log); err != nil {
+		return err
+	}
+	mgr.GetWebhookServer().Register("/v1/mutate", wh)
+
+	return nil
+}
+
+var _ admission.Handler = &mutationHandler{}
+
+type mutationHandler struct {
+	webhookHandler
+}
+
+// Handle the validation request
+func (h *mutationHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	log := log.WithValues("hookType", "mutation")
+	var timeStart = time.Now()
+
+	if isGkServiceAccount(req.AdmissionRequest.UserInfo) {
+		return admission.ValidationResponse(true, "Gatekeeper does not self-manage")
+	}
+
+	if req.AdmissionRequest.Operation != admissionv1beta1.Create &&
+		req.AdmissionRequest.Operation != admissionv1beta1.Update {
+		return admission.ValidationResponse(true, "Mutating only on create")
+	}
+
+	if h.isGatekeeperResource(ctx, req) {
+		return admission.ValidationResponse(true, "Not mutating gatekeeper resources")
+	}
+
+	requestResponse := mutationUnknownResponse
+	defer func() {
+		if h.reporter != nil {
+			if err := h.reporter.ReportMutationRequest(
+				requestResponse, time.Since(timeStart)); err != nil {
+				log.Error(err, "failed to report request")
+			}
+		}
+	}()
+
+	// namespace is excluded from webhook using config
+	if h.skipExcludedNamespace(req.AdmissionRequest.Namespace) {
+		requestResponse = mutationSkipResponse
+		return admission.ValidationResponse(true, "Namespace is set to be ignored by Gatekeeper config")
+	}
+
+	resp, err := h.mutateRequest(ctx, req)
+	if err != nil {
+		log.Error(err, "error executing query")
+		vResp := admission.Response{
```

Comment:
we can also use the library here:

https://github.com/kubernetes-sigs/controller-runtime/blob/f3831c24e5e99c242f993fde8aad18047ca47570/pkg/webhook/admission/response.go#L50-L61

Reply:
thanks for the tip
Done